SYCL/Plugin/interop-experimental-single-TU-SYCL-CUDA-compilation.cpp

```diff
@@ -0,0 +1,207 @@
+              test_cuda_function_3(a[i], b[i])) + //<-- __device__
+             test_cuda_function_4(a[i], b[i]) +   //<-- __host__ __device__
+             (test_cuda_function_5(a[i], b[i]) +  //<-- __device__
+              test_cuda_function_6(a[i], b[i]));  //<-- __device__
+    });
+  });
+
+  {
+    sycl::host_accessor c1{b_host, sycl::read_only};
+    sycl::host_accessor c2{b_dev, sycl::read_only};
+    for (size_t i = 0; i < n0; i++) {
+      // b_host = -1 b_dev
+      assert((c1[i] + c2[i] < 1e-5) && "Results mismatch!");
+    }
+  }
+}
+
+// ------------------------------------------------------------------------- //
+
+__device__ int test_cuda_function_1() {
+  return blockIdx.x * blockDim.x + threadIdx.x;
+}
+
+__global__ void test_cuda_kernel(int *out) {
+  int i = blockIdx.x * blockDim.x + threadIdx.x;
+  out[i] = i - test_cuda_function_1();
+}
+
+// Test CUDA kernel launch and CUDA API.
+void test_cuda_kernel_launch() {
+  // CUDA
+  const int n = 512;
+  std::vector<int> result(n, -1);
+  int *cuda_kern_result = NULL;
+
+  int block_size = 128;
+  dim3 dimBlock(block_size, 1, 1);
+  dim3 dimGrid(n / block_size, 1, 1);
+
+  cudaMalloc((void **)&cuda_kern_result, n * sizeof(int));
+
+  test_cuda_kernel<<<n / block_size, block_size>>>(cuda_kern_result);
+
+  cudaError_t error = cudaGetLastError();
+  if (error != cudaSuccess)
+    std::cerr << "CUDA ERROR: " << error << " " << cudaGetErrorString(error)
+              << std::endl;
+
+  cudaMemcpy(result.data(), cuda_kern_result, n * sizeof(int),
+             cudaMemcpyDeviceToHost);
+
+  for (size_t i = 0; i < n; i++)
+    assert((0 == result[i]) && "Kernel execution fail!");
+
+  cudaFree(cuda_kern_result);
+}
+
+// ------------------------------------------------------------------------- //
+
+__host__ float test_cuda_function_7() { return -1; }
+__device__ float test_cuda_function_7() { return 1; }
+
+__host__ float test_cuda_function_8() { return 3; }
+
+__device__ float test_cuda_function_9() { return 9; }
+
+int test_regular_function_0() { return test_cuda_function_7(); }
+
+int test_regular_function_1() { return test_cuda_function_8(); }
+
+int test_regular_function_2() { return test_cuda_function_9(); }
+
+// Test the correct emission of __device__/__host__ function when called by
+// regular functions.
+void test_regular_functions(sycl::queue &q) {
+
+  // regular func must returning the __host__ one (so, 1.0f)
+  assert((test_regular_function_0() == -1) &&
+         "Mismatch regular func to __host__");
+  assert((test_regular_function_1() == 3) &&
+         "Mismatch regular func to __host__");
+
+  sycl::buffer<int, 1> b_r{3};
+  q.submit([&](sycl::handler &h) {
+    sycl::accessor r{b_r, h, sycl::write_only};
+
+    h.single_task([=]() {
+      r[0] = test_regular_function_1(); //<-- points to __host__
+      r[1] = test_regular_function_2(); //<-- points to __device__
+      r[2] = test_regular_function_0(); //<-- points to __device__
```

Comment:
The order was confusing to me. I'd suggest going with 0, 1, 2 instead of 1, 2, 0.

Reply:
Sure, done.